flow/connectors/utils/cdc_records/cdc_records_storage.go

```diff
@@ -0,0 +1,167 @@
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/PeerDB-io/peer-flow/model"
+	util "github.com/PeerDB-io/peer-flow/utils"
+	"github.com/cockroachdb/pebble"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	/** begin with in-memory store, and then switch to Pebble DB
+		when the number of stored records crosses 1M
+	**/
+	defaultNumRecordsSwitchThreshold = 1_000_000
+)
+
+func encVal(val any) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	enc := gob.NewEncoder(buf)
+	err := enc.Encode(val)
+	if err != nil {
+		return []byte{}, fmt.Errorf("unable to encode value %v: %w", val, err)
+	}
+	return buf.Bytes(), nil
+}
+
+type cdcRecordsStore struct {
+	inMemoryRecords           map[model.TableWithPkey]model.Record
+	pebbleDB                  *pebble.DB
+	numRecords                int
+	dbFolderName              string
+	numRecordsSwitchThreshold int
+}
+
+func NewCDCRecordsStore(flowJobName string) *cdcRecordsStore {
+	return &cdcRecordsStore{
+		inMemoryRecords:           make(map[model.TableWithPkey]model.Record),
+		pebbleDB:                  nil,
+		numRecords:                0,
+		dbFolderName:              fmt.Sprintf("%s/%s_%s", os.TempDir(), flowJobName, util.RandomString(8)),
+		numRecordsSwitchThreshold: defaultNumRecordsSwitchThreshold,
+	}
+}
+
+func (c *cdcRecordsStore) initPebbleDB() error {
+	if c.pebbleDB != nil {
+		return nil
+	}
+
+	gob.Register(&model.InsertRecord{})
+	gob.Register(&model.UpdateRecord{})
+	gob.Register(&model.DeleteRecord{})
+	gob.Register(&time.Time{})
+	gob.Register(&big.Rat{})
+
+	var err error
+	// we don't want a WAL since cache, we don't want to overwrite another DB either
+	c.pebbleDB, err = pebble.Open(c.dbFolderName, &pebble.Options{
+		DisableWAL:         true,
+		ErrorIfExists:      true,
+		FormatMajorVersion: pebble.FormatNewest,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to initialize Pebble database: %w", err)
+	}
+	return nil
+}
+
+func (c *cdcRecordsStore) Set(key model.TableWithPkey, rec model.Record) error {
+	_, ok := c.inMemoryRecords[key]
+	if ok || len(c.inMemoryRecords) < defaultNumRecordsSwitchThreshold {
+		c.inMemoryRecords[key] = rec
+	} else {
+		if c.pebbleDB == nil {
+			logrus.Infof("more than %d primary keys read, spilling to disk", defaultNumRecordsSwitchThreshold)
+			err := c.initPebbleDB()
+			if err != nil {
+				return err
+			}
+		}
+
+		encodedKey, err := encVal(key)
+		if err != nil {
+			return err
+		}
+		// necessary to point pointer to interface so the interface is exposed
+		// instead of the underlying type
+		encodedRec, err := encVal(&rec)
+		if err != nil {
+			return err
+		}
+		// we're using Pebble as a cache, no need for durability here.
+		err = c.pebbleDB.Set(encodedKey, encodedRec, &pebble.WriteOptions{
+			Sync: false,
+		})
+		if err != nil {
+			return fmt.Errorf("unable to store value in Pebble: %w", err)
+		}
+	}
+	c.numRecords++
+	return nil
+}
+
+func (c *cdcRecordsStore) Get(key model.TableWithPkey) (model.Record, bool, error) {
```

Comment:
can you add a comment about the return value? say something like bool indicates found or something